### PR TITLE
Tests with matrices in echelon form

### PR DESCRIPTION
**Pull request by emmakatherina**
In July we discussed that we wanted to test our algorithm by creating random matrices from a matrix in reduced row echelon form. Until know I implemented a function that creates a matrix in reduced row echelon form and another function that generates an equivalent matrix. The equivalent matrix is to be used for testing the Chief and ChiefParallel algorithm and the goal is to complare the results to the original matrix in reduced row echelon form.

Unfortunately the original matrix does not have the same form as the echelon matrix obtained by Chief and ChiefParallel. I am still looking into that, but I would be happy to get some comments on what I have done so far and the way that I generate those matrices in reduced row echelon form.

@ssiccha @jnkbrachter 

Fixes issue #57, #40.

### CODE REVIEW

File: tst/parallel/echelon_form.tst [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+gap> echelon := echelonMat(dimension, dimension, rank, rs);;
+gap> shapeless := shapelessMat(echelon, dimension, dimension, rs, GF(q));;
+gap> result := ChiefParallel(GF(q), shapeless, numberChops, numberChops);;
+gap> result_std := EchelonMatTransformation(shapeless);;
+gap> -1 * result.vectors = result_std.vectors;
+true
+gap> -1 * result.coeffs = result_std.coeffs;
+true
+gap> -1 * result.vectors = MatrixByBlockMatrix(echelon);
+true
```
Comment (emmakatherina) [inline tst/parallel/echelon_form.tst]: The tests fail here!

File: tst/standard/echelon_form.tst [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+gap> echelon := echelonMat(dimension, dimension, rank, rs);;
+gap> shapeless := shapelessMat(echelon, dimension, dimension, rs, GF(q));;
+gap> result := Chief(GF(q), shapeless, numberChops, numberChops);;
+gap> result_std := EchelonMatTransformation(shapeless);;
+gap> -1 * result.vectors = result_std.vectors;
+true
+gap> -1 * result.coeffs = result_std.coeffs;
+true
+gap> -1 * result.vectors = MatrixByBlockMatrix(echelon);
+true
```
Comment (emmakatherina) [inline tst/standard/echelon_form.tst]: And here.

File: tst/parallel/echelon_form.tst (hunk 1 above)
Comment (emmakatherina) [reply tst/parallel/echelon_form.tst]: Okay, I jiust remembered that the rank is a separating invariant of equivalent matrices. So should we rather look at similar matrices? Can we compare the echelon forms then? Or do we not need to compare them anyway?

**Comment by ssiccha** [comment]
I've had a look at the new function and it looks reasonable. Two things looked odd though:
Apparently you used `MatrixByBlockMatrix` in an older version of your code and forgot to remove it again?

I gues you should pass a finite field to `RandomInvertibleMatrix`. Currently it constructs the matrices over the integers. Does `ChiefParallel` silently convert its input to a matrix over the finite field it was given? This could be the reason for your error message.

**Comment by emmakatherina** [comment]
Thank you, I changed the things you mentioned!

But the error is still there: For example when executing `tst/standard/echelon_form.tst` in line 14 the two variables contain the following:
```
Concatenation(result.coeffs, result.relations) * shapeless;
# Expected output:
# But found:
[ [ Z(5)^2, 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), Z(5)^2, Z(5)^2, 0*Z(5), Z(5)^3,
       Z(5)^2 ],
  [ 0*Z(5), Z(5)^2, 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), Z(5), Z(5)^3, Z(5)^3,
       Z(5)^3 ],
  [ 0*Z(5), 0*Z(5), Z(5)^2, 0*Z(5), 0*Z(5), Z(5)^2, Z(5)^0, Z(5)^2, Z(5)^3,
       Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), Z(5)^2, 0*Z(5), Z(5), Z(5), Z(5)^3, Z(5), Z(5)^0 ]
    , [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), Z(5)^2, Z(5)^0, Z(5)^0, 0*Z(5),
       0*Z(5), 0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
       0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
       0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
       0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
       0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
       0*Z(5) ] ]
########
########> Diff in tst/standard/echelon_form.tst:15
# Input is:
echelon;
# Expected output:
true
# But found:
[ [ Z(5)^0, 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), Z(5), Z(5)^2, Z(5)^3, Z(5)^2,
      0*Z(5) ],
  [ 0*Z(5), Z(5)^0, 0*Z(5), 0*Z(5), 0*Z(5), Z(5)^3, Z(5)^3, Z(5), Z(5)^3,
      Z(5)^3 ],
  [ 0*Z(5), 0*Z(5), Z(5)^0, 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), Z(5)^2,
      Z(5)^2 ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), Z(5)^0, 0*Z(5), 0*Z(5), Z(5)^0, Z(5)^2, Z(5)^0,
      Z(5)^0 ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), Z(5)^0, 0*Z(5), Z(5)^3, Z(5)^3, Z(5)^2,
      0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
      0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
      0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
      0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
      0*Z(5) ],
  [ 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5), 0*Z(5),
      0*Z(5) ] ]
########
```
I am still not quite sure how elements of finite fields are displayed, but at least gap thinks that those two matrices are not the same.

**Comment by emmakatherina** [comment]
I found a nicer way to look at it! :)
```
Display(Concatenation(result.coeffs, result.relations) * shapeless);
# Expected output:
# But found:
 4 . . . . 4 4 . 3 4
 . 4 . . . . 2 3 3 3
 . . 4 . . 4 1 4 3 2
 . . . 4 . 2 2 3 2 1
 . . . . 4 1 1 . . .
 . . . . . . . . . .
 . . . . . . . . . .
 . . . . . . . . . .
 . . . . . . . . . .
 . . . . . . . . . .
########
########> Diff in tst/standard/echelon_form.tst:15
# Input is:
Display(echelon);
# Expected output:
true
# But found:
 1 . . . . 2 4 3 4 .
 . 1 . . . 3 3 2 3 3
 . . 1 . . . . . 4 4
 . . . 1 . . 1 4 1 1
 . . . . 1 . 3 3 4 .
 . . . . . . . . . .
 . . . . . . . . . .
 . . . . . . . . . .
 . . . . . . . . . .
 . . . . . . . . . .
########
```

**Comment by emmakatherina** [comment]
Don't review yet. I just saw that I forgot to change something.

File: echelon_form.g [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    for i in [1 .. rank] do
+        for j in [1 .. (width-rank)] do
+            echelonMat[i][rank+j] := rightCorner[i][j];
+        od;
+    od;
+
+    return echelonMat;
+end;
+
+shapelessMat := function(mat, height, width, randomSeed, ring)
```
[the remark this answers is not in the document]
Comment (emmakatherina) [reply echelon_form.g]: Are we going to name all the "hidden" functions _GAUSS_xxx? Should we maybe rename those functions all at once?

File: echelon_form.g [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# This file creates a matrix in strict row echelon form to be used for testing and a function that gets a matrix and returns another matrix with the same strict row echelon form.
+
+echelonMat := function(height, width, rank, randomSeed, ring)
+   local echelonMat, rightCorner, i, j;
+
+    rightCorner := RandomMat(randomSeed, rank, width-rank, ring);
+    echelonMat := NullMat(height, width);
+
+    for i in [1 .. rank] do
+        echelonMat[i][i] := 1;
```
[the remark this answers is not in the document]
Comment (emmakatherina) [reply echelon_form.g]: I don't really know what you mean. If I write ``` echelonMat := One(ring); ``` then I have to delete 1's again and don't save anything, don't I?

File: echelon_form.g [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        for j in [1 .. (width-rank)] do
+            echelonMat[i][rank+j] := rightCorner[i][j];
+        od;
+    od;
+
+    return echelonMat;
+end;
+
+shapelessMat := function(mat, height, width, randomSeed, ring)
+    return RandomInvertibleMat(randomSeed, height, ring) * mat;
```
[the remark this answers is not in the document]
Comment (emmakatherina) [reply echelon_form.g]: Unfortunately I can only still decipher half of my notes about that. I wrote down `PseudoRandom` but this method returns only one element. Do we want to use Randomize? (https://www.gap-system.org/Manuals/pkg/cvec-2.6.0/doc/chap5.html)

File: echelon_form.g [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# This file creates a matrix in strict row echelon form to be used for testing and a function that gets a matrix and returns another matrix with the same strict row echelon form.
+
+echelonMat := function(height, width, rank, randomSeed, ring)
```
[the remark this answers is not in the document]
Comment (emmakatherina) [reply echelon_form.g]: Is there a specific way to create source comments? Do we want to use AutoDoc? https://www.gap-system.org/Manuals/pkg/AutoDoc/doc/chap2.html

File: echelon_form.g (hunk 5 above)
[the remark this answers is not in the document]
Comment (emmakatherina) [reply echelon_form.g]: Randomize seems to work only for compressed matrices in the filter IsGF2MatrixRep. I'm confused.

File: echelon_form.g [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    return echelonMat;
+end;
+
+shapelessMat := function(mat, height, width, randomSeed, ring)
+    local i, grp, left;
+
+    grp := GL(height, ring);
+    i := Length(GeneratorsOfGroup(grp));
+    Group_InitPseudoRandom(grp, i+3, 20);
```
Comment (ssiccha) [inline echelon_form.g]: Can you add a short source comment on why we do this?

Comment (emmakatherina) [reply echelon_form.g]: Done.

File: echelon_form.g [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    return echelonMat;
+end;
+
+_GAUSS_shapelessMat := function(mat, height, width, randomSeed, ring)
+    local i, grp, left;
+
+    # PseudoRandom is 'too random' for us and takes to much time to compute
+    # a matrix. That's why we set a PseudoRandomSeed beforehand, so that we
+    # can determine the degree of randomness.
```
Comment (ssiccha) [inline echelon_form.g]: There is no such thing as being 'too random'. ;) My suggestion:
```
    # PseudoRandom takes too much time to initialize PseudoRandomSeed for big values
    # of height. So we call the initialization ourselves with parameters that make it cheaper
    # to compute.
```

**Pull request merged**